stdlib/public/core/StringStorage.swift

```diff
@@ -16,59 +16,103 @@ import SwiftShims
 @usableFromInline
 internal class _AbstractStringStorage: __SwiftNativeNSString, _NSStringCore {
   // Abstract interface
-  internal var asString: String { get { Builtin.unreachable() } }
-  internal var count: Int { get { Builtin.unreachable() } }
-  internal func getOrComputeBreadcrumbs() -> _StringBreadcrumbs {
-    Builtin.unreachable()
-  }
+  internal var asGuts: _StringGuts { @_effects(readonly) get { Builtin.unreachable() } }
+  final internal var asString: String { @_effects(readonly) get { return String(asGuts) } }
+  internal var count: Int { @_effects(readonly) get { Builtin.unreachable() } }
 }
 
 // ObjC interfaces
 #if _runtime(_ObjC)
-extension _AbstractStringStorage {
-  @objc(length)
-  final internal var length: Int { return asString.utf16.count }
-
-  @objc(characterAtIndex:)
-  final internal func character(at offset: Int) -> UInt16 {
-    let str = asString
-    return str.utf16[str._toUTF16Index(offset)]
-  }
 
-  @objc(getCharacters:range:)
-  final internal func getCharacters(
+@inline(__always)
+@_effects(releasenone)
+private func _getCharacters<T:_AbstractStringStorage>(_ this:T,
```

Comment:
This used to be on _AbstractStringStorage, which meant it went through the dynamic version of asString. Making it generic and calling it from the subclasses lets us avoid that.